### ISSUE TITLE
Reproduction problems

### ISSUE THREAD

**Issue by JaniceXiong** [hfthair/emerald_crawler#5]
@memray @hfthair 
Hi, I try to train **BART-Facet I+C** on released dataset. But the settings in Appendix C can't run and get OOM error on 3090Ti 24GB. 
Your settings according to Appendix C Implementation Details. Is this right? :
- max-source-positions=10000
- max-tokens=10000
- update-freq=4
- total-num-update=20000
- lr=3e-0.5

So I change to setting like below and get very poor result (Rouge-L) using file2rouge: **Full/Purpose/Method/Findings/Value: 27.11/33.17/22.05/22.54/22.29**
- max-source-positions=4096
- max-tokens=4096
- update-freq=1
- total-num-update=80000
- lr=3e-0.5

I have some questions:
1. The aim of "extend the position embeddings to 10000 tokens" in Appendix C  is to take full text as input. If the input is I+C, What should the parameter **"max-source-positions"**  be set to？
2. What is the version of your fairseq? 
3. How many GPU devices you use on the V100?

Could you give some advice or release the detailed finetuning script like https://github.com/pytorch/fairseq/blob/master/examples/bart/README.summarization.md or upload trained model?
Thanks a lot!

**Comment by khushsi** [comment]
1. for I+C we used 4096 
3. only 1 gpu 

- we did the experiment on an aws server , which we truncated later  - so give me some time I will come back with more details by next thursday. and update on this comment

**Comment by JaniceXiong** [comment]
> * we did the experiment on an aws server , which we truncated later  - so give me some time I will come back with more details by next thursday. and update on this comment

Thanks for replying @khushsi ! So we use the same source length 4096 but get very different results. Hope to get more training details to reproduce the results soon and have a nice day :) !

**Comment by JaniceXiong** [comment]
@khushsi Hi, hope you have a nice day! Do you get any training details now?

**Comment by khushsi** [comment]
###  Fairseq version: 
Metadata-Version: 2.1
Name: fairseq
Version: 1.0.0a0+108f720

### Parameters while training
TOTAL_NUM_UPDATES=20000
WARMUP_UPDATES=500
LR=3e-05
MAX_TOKENS=4098
UPDATE_FREQ=4
BART_PATH=model.pt

sleep 1m

CUDA_VISIBLE_DEVICES=0 fairseq-train data-bin \
    --save-dir checkpoints \
    --restore-file checkpoints/model.pt \
    --max-tokens $MAX_TOKENS \
    --task translation \
    --batch-size 1 \
    --source-lang source --target-lang target \
    --truncate-source \
    --layernorm-embedding \
    --share-all-embeddings \
    --share-decoder-input-output-embed \
    --reset-optimizer --reset-dataloader --reset-meters \
    --required-batch-size-multiple 1 \
    --arch bart_large \
    --criterion label_smoothed_cross_entropy \
    --label-smoothing 0.1 \
    --dropout 0.1 --attention-dropout 0.1 \
    --weight-decay 0.01 --optimizer adam --adam-betas "(0.9, 0.999)" --adam-eps 1e-08 \
    --clip-norm 0.1 \
    --lr-scheduler polynomial_decay --lr $LR --total-num-update $TOTAL_NUM_UPDATES --warmup-updates $WARMUP_UPDATES \
    --fp16 --update-freq $UPDATE_FREQ \
    --skip-invalid-size-inputs-valid-test \
    --save-interval-updates 5000 \
    --max-update	$TOTAL_NUM_UPDATES \
    --find-unused-parameters;

**Comment by khushsi** [comment]
do you need output from our models - I have saved outputs from all our models -- ? 

But dont know if they would be of use

**Comment by khushsi** [comment]
also did you try with default max tokens ? how is your performance in that case ?

**Comment by JaniceXiong** [comment]
> ### Fairseq version:
> Metadata-Version: 2.1
> Name: fairseq
> Version: 1.0.0a0+108f720
> 
> ### Parameters while finetuning
> TOTAL_NUM_UPDATES=20000
> WARMUP_UPDATES=500
> LR=3e-05
> MAX_TOKENS=4098
> UPDATE_FREQ=4
> BART_PATH=model.pt
> 
> sleep 1m
> 
> CUDA_VISIBLE_DEVICES=0 fairseq-train data-bin
> --save-dir checkpoints
> --restore-file checkpoints/model.pt
> --max-tokens $MAX_TOKENS
> --task translation
> --batch-size 1
> --source-lang source --target-lang target
> --truncate-source
> --layernorm-embedding
> --share-all-embeddings
> --share-decoder-input-output-embed
> --reset-optimizer --reset-dataloader --reset-meters
> --required-batch-size-multiple 1
> --arch bart_large
> --criterion label_smoothed_cross_entropy
> --label-smoothing 0.1
> --dropout 0.1 --attention-dropout 0.1
> --weight-decay 0.01 --optimizer adam --adam-betas "(0.9, 0.999)" --adam-eps 1e-08
> --clip-norm 0.1
> --lr-scheduler polynomial_decay --lr $LR --total-num-update $TOTAL_NUM_UPDATES --warmup-updates $WARMUP_UPDATES
> --fp16 --update-freq $UPDATE_FREQ
> --skip-invalid-size-inputs-valid-test
> --save-interval-updates 5000
> --max-update $TOTAL_NUM_UPDATES
> --find-unused-parameters;

I think the most difference is that I set the **max-source-position** to 4096 because the average length of source input > 4000 tokens (this information is in notebook), but in your setting using the default max-source-position 1024. Because I enlarge the max-source-position, the number of training data decrease because the max-token in a batch is 4096 too.
It seems that increase the length of source input does not increase the performace of model.  Perhaps bart-large is unable to handle long input. I will try your settings again and I think **the saved output** are useful to do some analysis and comparisons. So if you can upload the outputs or trained model will be very helpful!
Thanks a lot for your detailed reply! @khushsi

**Comment by JaniceXiong** [comment]
Hi @khushsi 
I train the model using your parameters, but still got much lower Rouge results on test set. The Rouge-L Average_F is 27.77/33.74/22.41/23.62/22.82 in full/Purpose/Method/Findings/Value.

- Purpose:

Preparing documents... 0 line(s) ignored
Running ROUGE...
---------------------------------------------
1 ROUGE-1 Average_R: 0.60929 (95%-conf.int. 0.60440 - 0.61392)
1 ROUGE-1 Average_P: 0.39583 (95%-conf.int. 0.39204 - 0.39987)
1 ROUGE-1 Average_F: 0.45114 (95%-conf.int. 0.44789 - 0.45470)
---------------------------------------------
1 ROUGE-2 Average_R: 0.34336 (95%-conf.int. 0.33755 - 0.34891)
1 ROUGE-2 Average_P: 0.21286 (95%-conf.int. 0.20945 - 0.21617)
1 ROUGE-2 Average_F: 0.24674 (95%-conf.int. 0.24325 - 0.25025)
---------------------------------------------
1 ROUGE-L Average_R: 0.46714 (95%-conf.int. 0.46187 - 0.47249)
1 ROUGE-L Average_P: 0.29107 (95%-conf.int. 0.28784 - 0.29435)
1 ROUGE-L Average_F: **0.33744** (95%-conf.int. 0.33424 - 0.34090)

Elapsed time: 83.136 seconds

- Method:

Preparing documents... 0 line(s) ignored
Running ROUGE...
---------------------------------------------
1 ROUGE-1 Average_R: 0.44190 (95%-conf.int. 0.43821 - 0.44588)
1 ROUGE-1 Average_P: 0.30262 (95%-conf.int. 0.29924 - 0.30614)
1 ROUGE-1 Average_F: 0.33611 (95%-conf.int. 0.33321 - 0.33916)
---------------------------------------------
1 ROUGE-2 Average_R: 0.15043 (95%-conf.int. 0.14673 - 0.15423)
1 ROUGE-2 Average_P: 0.09896 (95%-conf.int. 0.09666 - 0.10138)
1 ROUGE-2 Average_F: 0.11134 (95%-conf.int. 0.10885 - 0.11385)
---------------------------------------------
1 ROUGE-L Average_R: 0.30249 (95%-conf.int. 0.29889 - 0.30629)
1 ROUGE-L Average_P: 0.19914 (95%-conf.int. 0.19668 - 0.20160)
1 ROUGE-L Average_F: **0.22413** (95%-conf.int. 0.22192 - 0.22649)

Elapsed time: 79.303 seconds

- Findings:

Preparing documents... 0 line(s) ignored
Running ROUGE...
---------------------------------------------
1 ROUGE-1 Average_R: 0.43887 (95%-conf.int. 0.43530 - 0.44267)
1 ROUGE-1 Average_P: 0.33822 (95%-conf.int. 0.33450 - 0.34213)
1 ROUGE-1 Average_F: 0.36069 (95%-conf.int. 0.35751 - 0.36386)
---------------------------------------------
1 ROUGE-2 Average_R: 0.14993 (95%-conf.int. 0.14627 - 0.15383)
1 ROUGE-2 Average_P: 0.11610 (95%-conf.int. 0.11293 - 0.11941)
1 ROUGE-2 Average_F: 0.12359 (95%-conf.int. 0.12042 - 0.12678)
---------------------------------------------
1 ROUGE-L Average_R: 0.29139 (95%-conf.int. 0.28802 - 0.29493)
1 ROUGE-L Average_P: 0.21971 (95%-conf.int. 0.21668 - 0.22272)
1 ROUGE-L Average_F: **0.23620** (95%-conf.int. 0.23342 - 0.23894)

Elapsed time: 94.428 seconds

- Value:

Preparing documents... 0 line(s) ignored
Running ROUGE...
---------------------------------------------
1 ROUGE-1 Average_R: 0.48026 (95%-conf.int. 0.47633 - 0.48436)
1 ROUGE-1 Average_P: 0.28517 (95%-conf.int. 0.28188 - 0.28845)
1 ROUGE-1 Average_F: 0.33687 (95%-conf.int. 0.33378 - 0.33967)
---------------------------------------------
1 ROUGE-2 Average_R: 0.16090 (95%-conf.int. 0.15687 - 0.16509)
1 ROUGE-2 Average_P: 0.09421 (95%-conf.int. 0.09180 - 0.09675)
1 ROUGE-2 Average_F: 0.11169 (95%-conf.int. 0.10898 - 0.11459)
---------------------------------------------
1 ROUGE-L Average_R: 0.33298 (95%-conf.int. 0.32930 - 0.33693)
1 ROUGE-L Average_P: 0.19074 (95%-conf.int. 0.18826 - 0.19318)
1 ROUGE-L Average_F: **0.22818** (95%-conf.int. 0.22571 - 0.23089)

Elapsed time: 79.595 seconds

- full:

Preparing documents... 0 line(s) ignored
Running ROUGE...
---------------------------------------------
1 ROUGE-1 Average_R: 0.59983 (95%-conf.int. 0.59737 - 0.60246)
1 ROUGE-1 Average_P: 0.41722 (95%-conf.int. 0.41444 - 0.41998)
1 ROUGE-1 Average_F: 0.48088 (95%-conf.int. 0.47881 - 0.48310)
---------------------------------------------
1 ROUGE-2 Average_R: 0.25406 (95%-conf.int. 0.25118 - 0.25709)
1 ROUGE-2 Average_P: 0.17483 (95%-conf.int. 0.17283 - 0.17702)
1 ROUGE-2 Average_F: 0.20241 (95%-conf.int. 0.20022 - 0.20474)
---------------------------------------------
1 ROUGE-L Average_R: 0.35071 (95%-conf.int. 0.34819 - 0.35337)
1 ROUGE-L Average_P: 0.23919 (95%-conf.int. 0.23750 - 0.24091)
1 ROUGE-L Average_F: **0.27772** (95%-conf.int. 0.27600 - 0.27948)

Elapsed time: 522.854 seconds

Do you have any idea about that?

**Comment by khushsi** [comment]
did you check with different checkpoints -- is your rouge score still increasing -- ? Can you send me that trend 
so this is with 1024 source tokens right ? 
the first output is for BART - FACET right ? 
how did you get the data ? can I see the data statistics

**Comment by JaniceXiong** [comment]
@khushsi Hi :)
- the source token is default 1024
- the training loss trend is like below, I use tensorboardX in fairseq to get this
![image](https://user-images.githubusercontent.com/43976565/132532711-a7c4412a-eefd-4f61-ae96-bd55e657f3d7.png)
- For data, I use the train.urls, dev.urls, test.urls in your repo https://github.com/hfthair/emerald_crawler/tree/main/data. The statistic look like below, the first one is the statistics in your notebook, the latter is the statistic in my notebook using the same released code.
![image](https://user-images.githubusercontent.com/43976565/132533926-301780d4-0c21-4ee8-8a73-3ee6a0569513.png)
![image](https://user-images.githubusercontent.com/43976565/132533964-9ec19799-a236-485e-a265-7e3b6630361c.png)
And the BART-FACET input looks like
![image](https://user-images.githubusercontent.com/43976565/132534139-67398740-a1a6-4d2d-b53a-25249f0abca1.png)

- I only check the results of checkpoint_best

**Comment by JaniceXiong** [comment]
> @khushsi Hi :)
> 
> * the source token is default 1024
> * the training loss trend is like below, I use tensorboardX in fairseq to get this
>   ![image](https://user-images.githubusercontent.com/43976565/132532711-a7c4412a-eefd-4f61-ae96-bd55e657f3d7.png)
> * For data, I use the train.urls, dev.urls, test.urls in your repo https://github.com/hfthair/emerald_crawler/tree/main/data. The statistic look like below, the first one is the statistics in your notebook, the latter is the statistic in my notebook using the same released code.
>   ![image](https://user-images.githubusercontent.com/43976565/132533926-301780d4-0c21-4ee8-8a73-3ee6a0569513.png)
>   ![image](https://user-images.githubusercontent.com/43976565/132533964-9ec19799-a236-485e-a265-7e3b6630361c.png)
>   And the BART-FACET input looks like
>   ![image](https://user-images.githubusercontent.com/43976565/132534498-6f554dc6-a4d4-47c7-b3b4-e3144cd35dfa.png)
> * I only check the results of checkpoint_best
> 
> Perhaps you can upload saved outputs or evaluate scripts to keep all thing the same?

@khushsi  Do you set a particular seed or something?

**Comment by khushsi** [comment]
Hi Xiong,

I will upload code/model output/my scripts  to repository but have some deadlines so give me this weekend. 
[link](https://github.com/khushsi/Finetuning_BART_for_FACET_Summarization)
Hopefully than we can figure out what is the difference. But you need to be a little patient / sorry for I have lots of things on my hand currently.

Regards
Khushboo

**Comment by JaniceXiong** [comment]
> I will upload code/model output/my scripts to repository but have some deadlines so give me this weekend.
> [link](https://github.com/khushsi/Finetuning_BART_for_FACET_Summarization)
> Hopefully than we can figure out what is the difference. But you need to be a little patient / sorry for I have lots of things on my hand currently.
> 
> Regards
> Khushboo

Wow, thanks a lot!

**Comment by khushsi** [comment]
Hi, 
I have uploaded the output andy scripts. Let me know if you  need anything else.

**Comment by JaniceXiong** [comment]
> MAX_TOKENS=4098

@khushsi  here you given parameters is 4098 but in [new repo ](https://github.com/khushsi/Finetuning_BART_for_FACET_Summarization/blob/main/scripts/emerald_train_finetune.sh)  the same parameter is 1024. What is the parameters you exactly use in finetuning?

**Comment by JaniceXiong** [comment]
> BART_PATH=model.pt

And the model.pt here is bart.base or bart.large in https://github.com/pytorch/fairseq/tree/main/examples/bart?

**Comment by khushsi** [comment]
So it depends on the model.. if I train for 10k max tokens it's 10k
Other wise default 1k

For bartfacet oa test we used 1k only

Oh I will provide you with the code change for max tokens which I have done
in fairseq

**Comment by khushsi** [comment]
bart.large

**Comment by JaniceXiong** [comment]
> So it depends on the model.. if I train for 10k max tokens it's 10k

sorry, I didn't understand what 10k and 1k means here. 
So I need to finetune bart_large model using the parameters https://github.com/khushsi/Finetuning_BART_for_FACET_Summarization/blob/main/scripts/emerald_train_finetune.sh and test the model on test_oa dataset. Perhaps tomorrow I can get the results.

**Comment by khushsi** [comment]
so initially we experimented with allowing more than default tokens - i have uploaded the code in script/train.py  but you dont have to bother as for this model I only used 1024

ok hopefully now you will get inline results
